src/CppUTest/TestFailure.cpp

```diff
@@ -190,7 +190,25 @@ DoublesEqualFailure::DoublesEqualFailure(UtestShell* test, const char* fileName,
 {
     message_ = createUserText(text);
 
-    message_ += createButWasString(StringFrom(expected, 7), StringFrom(actual, 7));
+    bool nums_valid = !((PlatformSpecificIsNan(expected)) || (PlatformSpecificIsInf(expected)) ||
+                        (PlatformSpecificIsNan(actual)) || (PlatformSpecificIsInf(actual)));
+    int prec = 3;
+    SimpleString sexp = StringFrom(expected, prec);
+    SimpleString sact = StringFrom(actual, prec);
+    int dsize = sizeof(double);
+    const int prec_max = dsize == 8 ? 17 : 7;
+
+    if (nums_valid)
+    {
+        while ((prec < prec_max) && (sexp == sact))
+        {
+            prec++;
+            sexp = StringFrom(expected, prec);
+            sact = StringFrom(actual, prec);
+        }
+    }
+    message_ += createButWasString(sexp, sact);
+
```

Comment:
I'm looking at these few lines and I can not figure out how this code may be transferred into the SimpleString class. 
DoublesEqualFailure class is supposed to create verbose diagnostic output - and that's what it's doing! 
String is creating formatted double value output when asked, but it has nothing to do with formatting the output, according to how the other String looks like. It's this Failure class job to specify the format precision to achieve valuable message. Am I wrong?

Checking the double values validity is ugly, yet necessary.
Now all I can propose it this:

```C++
    const bool nums_valid = !((PlatformSpecificIsNan(expected)) || (PlatformSpecificIsInf(expected)) ||
                              (PlatformSpecificIsNan(actual)) || (PlatformSpecificIsInf(actual)));
    int prec = 3;
    const int dsize = sizeof(double);
    const int prec_max = dsize == 8 ? 17 : 7;
    SimpleString sexp, sact;

    if (nums_valid)
    {
        while ((prec < prec_max) && (sexp == sact))
        {
            sexp = StringFrom(expected, prec);
            sact = StringFrom(actual, prec);
            prec++;
        }
    }
    else
    {
        sexp = StringFrom(expected);
        sact = StringFrom(actual);
    }
    message_ += createButWasString(sexp, sact);
```